profiling/src/collections/string_table/mod.rs

```diff
@@ -0,0 +1,279 @@
+        // capacity after the first insertion is quite small, as in 3. This is
+        // a bit too small and there are frequent reallocations. For one sample
+        // with endpoint + code hotspots, we'd have at least these strings:
+        // - ""
+        // - At least one sample type
+        // - At least one sample unit--already at 3 without any samples.
+        // - "local root span id"
+        // - "span id"
+        // - "trace endpoint"
+        // - A file and/or function name per frame.
+        // So with a capacity like 3, we end up reallocating a bunch on or
+        // before the very first sample. The number here is not fine-tuned,
+        // just skipping some obviously bad, tiny sizes.
+        strings.reserve(32);
+
+        // Always hold the empty string as item 0. Do not insert it via intern
+        // because that will try to allocate zero-bytes from the storage,
+        // which is sketchy.
+        strings.insert("");
+
+        Self { bytes, strings }
+    }
+
+    /// Returns the number of strings currently held in the string table.
+    #[inline]
+    #[allow(clippy::len_without_is_empty)]
+    pub fn len(&self) -> usize {
+        self.strings.len()
+    }
+
+    /// Adds the string to the string table if it isn't present already, and
+    /// returns a [StringId] that corresponds to the order that this string
+    /// was originally inserted.
+    ///
+    /// # Panics
+    /// This panics if the allocator fails to allocate a new chunk/node.
+    pub fn intern(&mut self, str: &str) -> StringId {
+        let set = &mut self.strings;
+        match set.get_index_of(str) {
+            Some(offset) => StringId::from_offset(offset),
+            None => {
+                // No match. Get the current size of the table, which
+                // corresponds to the StringId it will have when inserted.
+                let string_id = StringId::from_offset(set.len());
+
+                // Make a new string in the arena, and fudge its lifetime
+                // to appease the borrow checker.
+                let new_str = {
+                    // PANIC: the intern API doesn't allow for failure, so if
+                    // this allocation fails, panic. The current
+                    // implementation of `ChainAllocator` will fail if the
+                    // underlying allocator fails when asking for a new chunk.
+                    // This is expected to be rare.
+                    let s = ArenaAllocator::allocate(&self.bytes, str)
+                        .expect("allocator for StringTable::intern to succeed");
+
+                    // SAFETY: all references to this value get re-narrowed to
+                    // the lifetime of the string table or iterator when
+                    // exposed to the user. The string table and iterator will
+                    // keep the arena alive, making the access safe.
+                    unsafe { core::mem::transmute::<&str, &'static str>(s) }
```

Comment:
is it possible to give it the same lifetime as `self`?

Reply:
There is no such thing as a `'self` lifetime, unfortunately 🙁